### ISSUE TITLE
CA1812 (Avoid uninstantiated internal classes) doesn't account for subtypes being instantiated

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-analyzers#1370]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

AvoidUninstantiatedInternalClassesAnalyzer

#### Repro steps

```
internal class A
{
    public A(int x)
    {
    }
}

internal class B : A
{
    public B(int x): base (x)
    {
    }
}

internal class C
{
    public C()
    {
    }
}

internal class D : C
{
    public D()
    {
    }

    static void M()
    {
        var x = new B(0);
        var y = new D();
    }
}
```

#### Expected behavior

No CA1812 on any of the types

#### Actual behavior

CA1812 is reported for `A` and `C`

**Comment by mavasani** [comment]
This diagnostic is enabled by default in VSIX and leading to quite a few false reports on real world solutions.

**Comment by mavasani** [comment]
We should probably just skip any type which has a subtype in the compilation to reduce noise.

**Comment by mavasani** [comment]
Fact that diagnostic is reported on `A` and analyzer is registering for all object creation expressions likely indicates an IOperation bug that we are not making callbacks for executable code in ctor initializers, which needs a corresponding bug to be filed in Roslyn.
